src/quest/resources.py

```diff
@@ -0,0 +1,114 @@
+    def __init__(self):
+        self._resource_streams: dict[str, set[ResourceStreamManager.ResourceStream]] = {}
+
+    class ResourceStream:
+        def __init__(self,
+                     get_resources: Callable[[], Coroutine],
+                     is_workflow_complete: Callable[[], bool],
+                     on_open: Callable[['ResourceStreamManager.ResourceStream'], None],
+                     on_close: Callable[['ResourceStreamManager.ResourceStream'], None]
+                     ):
+            self._stream_gate = asyncio.Event()
+            self._update_event = asyncio.Event()
+            self._is_entered = False
+
+            self._get_resources = get_resources
+            self._is_workflow_complete = is_workflow_complete
+            self._on_open = on_open
+            self._on_close = on_close
+
+        def __enter__(self):
+            self._is_entered = True
+            self._on_open(self)
+            logging.debug(f'Resource stream opened for {id(self)}')
+            return self
+
+        def __exit__(self, exc_type, exc_value, traceback):
+            self._is_entered = False
+            self._on_close(self)
+            logging.debug(f'Resource stream closed for {id(self)}')
+
+        async def __aiter__(self):
+            """
+            Provide a stream of resource snapshots for this workflow.
+            Everytime the workflow resource state changes, an update will be published.
+
+            NOTE: Once you start the resource stream,
+            the workflow will not progress unless you iterate this stream or exit the `with` context.
+            """
+
+            if not self._is_entered:
+                raise Exception('ResourceStream must be used in a `with` context.')
+
+            # TODO: Do we still need this explanation?
+            # The caller of this function lives outside the step management of the historian
+            #         -> don't replay, just yield event changes in realtime
+
+            yield await self._get_resources()  # Yield the current resources immediately
+
+            # TODO: What if the workflow is complete but we are already in this loop awaiting a update_event? How close?
+            # Yield new resources updates as they become available
+            while not self._is_workflow_complete():
+                await self._update_event.wait()
+                self._update_event.clear()
+                yield await self._get_resources()
+                self._stream_gate.set()
+
+    def _on_open(self, identity, res_stream: ResourceStream):
+        if identity not in self._resource_streams:
+            self._resource_streams[identity] = set()
+
+        self._resource_streams[identity].add(res_stream)
+
+    def _on_close(self, identity, res_stream: ResourceStream):
+        res_stream._stream_gate.set()
+        self._resource_streams[identity].remove(res_stream)
+
+        if not self._resource_streams.get(identity):  # Clean up dictionary values if needed
```

Comment:
`if not self._resource_streams[identity]`

In this situation, we are guaranteed to have `identity` in `_resource_streams`, so this logic will work. But if there was a chance `identity` was not in `_resource_streams`, using `get` will return `None`, which will be falsy and lead to an error when we try to pop identity. 

So, to avoid confusion (is `get` returning `None`?) let's use the bracket operator.